metricbeat/module/kubernetes/statemetrics/_meta/docs.asciidoc

```diff
@@ -0,0 +1,94 @@
+The `statemetrics` metricset for Kubernetes module fetches metrics from `kube-state-metrics` prometheus endpoint and pushes formated events to the configured output.
+
+It has been tested using `kube-state-metrics` `v1.7.0`. When tested with other versions it should still work with all unchanged metrics while ignoring those that do not match.
+
+[float]
+=== Configuration
+
+When configuring this metricset make sure that you have permissions to reach the `kube-state-metrics` endpoint.
+
+```yaml
+# State metrics from kube-state-metrics service:
+- module: kubernetes
+  enabled: true
+  metricsets:
+    - statemetrics
+  period: 50s
+  hosts: [http://kube-state-metrics.kube-system:8080]
+```
+
+[float]
+=== Filtering
+
+By default all groups data is retrieved. Due to the considerable ammount of metrics retrieved filtering is recommended when not using all of the data retrieved. Filtering can be specified pre-process and post-process.
+Post-process filtering is done via the `drop_event` processors. Pre-process filtering on the other hand is managed at the metricset configuration.
+
+This metricset data is classified in the following groups, as defined at `kube-state-metrics`:
+
+- `certificatessigningrequest`
+- `configmap`
+- `cronjob`
+- `daemonset`
+- `deployment`
+- `endpoint`
+- `horizontalpodautoscaler`
+- `ingress`
+- `job`
+- `limitrange`
+- `namespace`
+- `node`
+- `persistentvolume`
+- `persistentvolumeclaim`
+- `pod`
+- `poddisruptionbudget`
+- `replicaset`
+- `replicationcontroller`
+- `resourcequota`
+- `secret`
+- `service`
+- `statefulset`
+- `storageclass`
+- `verticalpodautoscaler`
+
+We use the group names above to specify pre-process filtering. You can choose one of blacklist or whitelist filter strategies.
+When using blacklabel, all groups but the ones specified in the filter list will be added:
+
+[source,yaml]
+----
+# State metrics from kube-state-metrics service:
+- module: kubernetes
+  enabled: true
+  filter:
+    blacklist:
```

Comment:
what would be the reasoning to do blacklist instead of `drop_event`?

Reply:
**Pros**

blacklist happens before processing, it avoids including items in the mapping object, should save memory + CPU + time.
In fact, I was wondering how to introduce this same feature for all prometheus metrics, but don't want to abstract it until that is proved to be worth it for this metricset.

for me it feels better to write configuration to receive everything but `configmaps` and `damonsets` like:

```
metricset:
  kube-state-metrics:
    i-want-all-but:
      - configmaps
      - daemonsets
```

than 
```
metricset:
  kube-state-metrics:

processors:
  once-processed-remove:
    - events-that-match
       event.somekey equals kube_configuration_*
    - events-that-match
       event.somekey equals kube_daemonset_*
```

**Cons**

not using `drop_event` is not beats standard, this might be a one off unless we find this valuable for other metrics. That said, you can still use `drop_event` even if blacklisting makes it in this PR

Reply:
I wonder how often people will want to disable these. What would be the expected behavior for resources that are not present in k8s? This will probably report none or only a few documents.

Reply:
I agree on doubts when introducing features no one asked for

I think there is a total of 28 groups, among them `replicationcontrollers`, `signingcertificaterequests`, `storageclasses` ... and the default would be enabling them all. But as soon as a user notices all that information that is not useful for their use cases, or that might bring redundant information (sometimes `pv`/`pvc`), and that keeps on using storage, it looks like a candidate to be filtered.

> What would be the expected behavior for resources that are not present in k8s?

not present at prometheus --> no data is processed nor sent
present and boring (storageclasses, whatever the user considers boring) --> constant data with no value
The use case for this, it would be the same as the `drop_event` processor, but other than saving storage at ES, it saves resources + time, and groups using _familiar_ names

Reply:
I'm feeling like dropping blacklisting and whitelisting filtering. Those are a lot welcomed here, but don't want to go forward until discussing out of band with the team.

Reply:
SGTM, we can always add them in a followup PR